File: apps/quickink/android/app/src/main/java/app/quickink/mobile/data/location/LocationEntity.kt
/*
 * LocationEntity.kt
 *
 * Room @Entity for the `locations` table — user-defined places
 * ("Home", "Work", "Cafe", etc.) that captures can optionally be
 * tagged with. Many-to-many to captures via `capture_locations`,
 * mirroring the `tags` ↔ `capture_tags` shape.
 *
 * Seeded with "Home" and "Work" on first launch (see
 * LocationRepository.seedDefaultsIfEmpty).
 */

package app.quickink.mobile.data.location

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Index
import androidx.room.PrimaryKey

@Entity(
    tableName = "locations",
    indices = [
        Index(value = ["user_id", "name"], unique = true),
        Index(value = ["user_id", "position"], name = "idx_locations_user_position"),
        Index(value = ["dirty"], name = "idx_locations_dirty"),
        Index(value = ["deleted_at"], name = "idx_locations_tombstone"),
    ],
)
data class LocationEntity(
    @PrimaryKey
    @ColumnInfo(name = "id")
    val id: String,

    @ColumnInfo(name = "user_id")
    val userId: String,

    @ColumnInfo(name = "name")
    val name: String,

    @ColumnInfo(name = "position", defaultValue = "0")
    val position: Int,

    /**
     * Optional hex color for the location chip (e.g. "#E66943"). NULL →
     * UI falls back to the accent tint.
     */
    @ColumnInfo(name = "color")
    val color: String? = null,

    /**
     * Coordinates + formatted address. All three are NULL for
     * locations whose physical place hasn't been set yet (the
     * seeded "Home" / "Work" rows, or a freshly-typed name).
     * Populated by the location editor via "Use current location"
     * (GPS + reverse geocode) or "Search address" (forward geocode).
     */
    @ColumnInfo(name = "latitude")
    val latitude: Double? = null,

    @ColumnInfo(name = "longitude")
    val longitude: Double? = null,

    @ColumnInfo(name = "address")
    val address: String? = null,

    @ColumnInfo(name = "drive_file_id")
    val driveFileId: String?,

    @ColumnInfo(name = "created_at")
    val createdAt: String,

    @ColumnInfo(name = "updated_at")
    val updatedAt: String,

    @ColumnInfo(name = "dirty", defaultValue = "1")
    val dirty: Boolean,

    @ColumnInfo(name = "deleted_at")
    val deletedAt: String?,
)

File: apps/quickink/ios/QuickInk/Data/LocationEntity.swift
/*
 * LocationEntity.swift
 *
 * GRDB record for the `locations` table — user-defined places ("Home",
 * "Work", "Cafe") that captures can optionally be tagged with.
 * Many-to-many to captures via `capture_locations`. Mirror of
 * Android's `LocationEntity.kt`.
 *
 * Surfaced as "Places" in the Home screen UI; the table + sync wire
 * format keep the legacy `location` kind for cross-platform parity.
 */

import Foundation
import GRDB

public struct LocationEntity: Codable, FetchableRecord, PersistableRecord, Equatable, Sendable, Identifiable {

    public static let databaseTableName = "locations"

    public var id: String
    public var userId: String
    public var name: String
    public var position: Int
    public var color: String?
    public var latitude: Double?
    public var longitude: Double?
    public var address: String?
    public var driveFileId: String?
    public var createdAt: String
    public var updatedAt: String
    public var dirty: Bool
    public var deletedAt: String?

    public init(
        id: String,
        userId: String,
        name: String,
        position: Int,
        color: String? = nil,
        latitude: Double? = nil,
        longitude: Double? = nil,
        address: String? = nil,
        driveFileId: String? = nil,
        createdAt: String,
        updatedAt: String,
        dirty: Bool = true,
        deletedAt: String? = nil
    ) {
        self.id          = id
        self.userId      = userId
        self.name        = name
        self.position    = position
        self.color       = color
        self.latitude    = latitude
        self.longitude   = longitude
        self.address     = address
        self.driveFileId = driveFileId
        self.createdAt   = createdAt
        self.updatedAt   = updatedAt
        self.dirty       = dirty
        self.deletedAt   = deletedAt
    }

    public enum CodingKeys: String, CodingKey {
        case id
        case userId      = "user_id"
        case name
        case position
        case color
        case latitude
        case longitude
        case address
        case driveFileId = "drive_file_id"
        case createdAt   = "created_at"
        case updatedAt   = "updated_at"
        case dirty
        case deletedAt   = "deleted_at"
    }
}
